File: solution/kotlin/week05-exercise2/src/main/kotlin/command/CommandProcessor.kt
package command

class CommandProcessor {
    private val commands: MutableMap<String, Command> = mutableMapOf()

    val commandsNames: Set<String>
        get() = commands.keys

    fun addCommand(commandName: String, command: Command) {
        commands[commandName] = command
    }

    fun processCommand(commandName: String, printTo: (String) -> Unit) {
        val cmd = commands[commandName] ?: Command { "Command not recognized." }
        printTo(cmd.executeAndDisplayResult())
    }
}

File: solution/c#/week05-complexity/Exercise2CommandProcessor/CommandProcessor.cs
namespace Exercise2CommandProcessor
{
    public class CommandProcessor
    {
        private readonly Dictionary<string, ICommand> commands = [];

        public IEnumerable<string> GetCommandsNames()
    {
        return commands.Keys;
    }

        public void AddCommand(string commandName, ICommand command)
    {
        commands[commandName] = command;
    }

        public void ProcessCommand(string commandName, Action<string> printTo)
    {
        printTo(
            commands.TryGetValue(commandName, out var cmd)
                ? cmd.ExecuteAndDisplayResult()
                : "Command not recognized.");
    }
    }
}